hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ReplicationManager.java

```diff
@@ -325,10 +329,22 @@ public synchronized void processAll() {
         containerManager.getContainers();
     ReplicationManagerReport report = new ReplicationManagerReport();
     ReplicationQueue newRepQueue = new ReplicationQueue();
+    Map<ContainerID, MoveDataNodePair> pendingMoves =
+        moveManaer.getPendingMove();
```

Comment:
Does the ReplicationManager really need to know about the pending moves?

RM is spilt into two parts. The first, is finding containers with a problem (this bit), containers with a problem get added to a queue which is then processed later. Even if the container was identified as over-replicated here, MoveManager would have scheduled a delete of the required container before it is processed on the second queue.

Reply:
Thinking about this some more - what if the over-replication processing was "space aware". Given two replicas, always remove the one on the host with the least free space.

The balancer will always be adding containers to hosts with more free space, and removing replicas from hosts with less free space, so it seems that the standard over-replication handling could take care of this if it considered the free space on host when deciding which to remove.

Reply:
>Does the ReplicationManager really need to know about the pending moves?

in vast majority of cases, no need. but i take two corner cases into account. 
1 if a ratis container has only two replicas, r1(dn1) , r2(dn2) , now it is under-replicated. if we schedule a move for r1 and everything goes well, there will be a r3 in dn3 ultimately and this container is not under-replicated again.
if RM is not aware of this move, it will schedule to replication to a randomly selected dn4 to handle this under-replicated state . after the replication is done, the container is over-replicated(4 replicas), another deletion will be scheduled. so if RM is aware of this move, the deletion can be avoided.

2 if a container is over-replicated(r1, r2, r3, r4), and we want to move r1 to dn5, then movemanager will send a replication command to dn5. if RM is not ware of this move and it find this over-replicated container, it may send a deletion command to r1, which will fail the move. 

>what if the over-replication processing was "space aware"

i have a [ jira ](https://issues.apache.org/jira/browse/HDDS-5278) to talk about this.
beside what @lokesh said, another reason is scm leader switch. let`s say a container has r1 , r2, r3, r4 on dn1, dn2, dn3, dn4. if rm1 in scm1(leader) find this container is over-replicated and r1 has the least free space, it will send a deletion to dn1. scm1 steps down and scm2 becomes the leader. if some new data has been writen to dn3 and it gets the least free space now, rm2 in scm2 will send a deletion command to dn3 to delete r3. now , two datanodes receives the deletion command while only one replica need to be deleted.

this [jira ](https://issues.apache.org/jira/browse/HDDS-4589) make sure all the scms will delete replicas in a certain sequence.

Reply:
A fundamental design problem with the old RM was it placing many hours or even days worth or commands on the DNs without anyway to see what progress they are making.

It seems the balancer has make this same mistake, in that it does not limit what is queued on a DN. Ideally, the balancer should schedule no more work than can be done in a heartbeat or two. If we have under utilized nodes, they are going to be the target for replicate commands, so the limit for the number queued should be quite small per heartbeat. The delete commands might be more scattered, but we should even limit them, but to a higher number.

If we avoid deep queues on the DNs, then we also avoid the problem with a leader switch on SCM.

At the moment ReplicationManager does not replicate its pending ops or commands to each SCM. The goal is to make the queue on the DN small, so that even if a leader switch did happen, there is not much more work to do, otherwise RM can fall into this same problem or scheduling multiple deletes.

Better than that, we could also make the DN reject certain commands that don't have the current SCM Leader term. For example, any replicate, delete container, reconstruct container can be discarded if the SCM term has increased and the replication manager will just re-create commands on its next pass.

There isn't really any reason that the Balancer could not do the same. Ie, schedule work only on DNs with the capacity to process it, and then drop all commands on a leader switch. It would make the balancer a lot more simple overall.

Then, we could have the over-replication handling try to take space into account and remove the replicas from the most over-used nodes, making it do the correct thing more of the time. In the event of a failover, some over-replicated containers caused by the balancer might have the wrong replica removed, but if the free space is so finely balanced between the nodes that it removes the wrong one, it doesn't really matter.

Reply:
> 1 if a ratis container has only two replicas, r1(dn1) , r2(dn2) , now it is under-replicated. if we schedule a move for r1 and everything goes well, there will be a r3 in dn3 ultimately and this container is not under-replicated again.

RM is aware of this via the pending ops. It will see the container as under-replicated, but with pending ops it will see it as OK and will not schedule anything further.

> if a container is over-replicated(r1, r2, r3, r4), and we want to move r1 to dn5, then movemanager will send a replication command to dn5. if RM is not ware of this move and it find this over-replicated container, it may send a deletion command to r1, which will fail the move.

This example is slightly strange - what you are saying could happen, but really rather than "moving" a container from the over-used node to a lesser used node, we should just delete it from the over-used node.

Also, ideally, a replicate command for a move could have any of the source replicas as the place to read from. If you only have a few over used nodes on the cluster, then they will be a bottleneck for copying from. You don't have to read from the source you are going to delete. You can read from any source. Infact, I think the replicate command lets you pass as many sources as you want, and then it can retry if the copy fails from one source.